### ISSUE TITLE
Clean up editorID vs editor declaredClass

### ISSUE THREAD

**Issue by JonFerraiolo** [maqetta/maqetta#1872]
What I remember from looking at the code when trying to determine which editor we are in:
- Sometimes we check editorID and sometimes the editor's declaredClass
- There seemed to be two different values used for editorID. I don't remember the details, but I think it was the review editor that had two different strings used in two different places. Maybe theme editor and page editor did this also.
- It looked to me like editor's declaredClass was a reliable indicator against which you could make a reliable comparison

We need to search through the code and clean all of this up.

**Comment by childsb** [comment]
A note on this--

The editors inherit a method supports("something").  You can use this method to determine if a specific function is supported by the active editor then hide/show/do something based on that.

I think supports(...) is cleaner than checking a specific declared class.  That way the views dont have to be aware of which editors actually support them, they can just do if(editor.supports("properties")) show(); And its easier to add new editors.

ex:

supports : function(something){
    return something=="properties" || something=="comments" || something=="states"

}

shows an editor supports properties, comments and states.

**Comment by JonFerraiolo** [comment]
yes, supports() is way better than hard-coding a class name or a particular ID string. It helps deal with the possibility that we might have object-oriented hierarchies of editor classes (which in fact we have, where we have inherited classes: Editor->ModelEditor->PageEditor->VisualPageEditor. 

Maybe we should have a convention where we have an _supports array that base classes can add to in their constructor and the derived classes can push into that array? Just thinking ahead.

**Comment by waynevicknair** [comment]
the 'supports()' discussion seems orthogonal to identifying an editor. in the places where we test editorID or declaredClass we are trying to determine which _editor_ not which _features_ are present. the declaredClass is definitely the wrong thing to check as this closes the door on OO by introducing a hard dependency. we spent the effort to gain some inversion of control via the plugin metadata; we should use the editorID for now.

**Comment by childsb** [comment]
@waynevicknair why would we want to identify an editor besides determining to turn on or off a specific feature/code path?

Using editor ID and declaredClass are basically the same thing.  We need to abstract the functionality based conditionally on something other than a 1:1 identity (editorID or declaredClass).  

As an example, the theme editor doesn't support cascade functionality.  If we check editor Id to disable cascade, then every other editor that doesn't support cascade will either have to be added to the check or have the same editor ID as the theme editor.

Similar with a problems view, wizards etc..  They are a unit of function that one or more editors support, and we DON'T want to tie them to a specific editor ID or class.

**Comment by JonFerraiolo** [comment]
I think Brad's right. It seems to me we could use supports() everywhere we use declaredClass or editorID. The downside is that we will have a long list of strings that we regularly change within all of the supports() functions.

**Comment by JonFerraiolo** [comment]
@billreed63 

OK, I'm flip-flopping on this issue after doing an inventory (see below). My conclusion is that attempting to use `supports()` would be a huge waste of time, and instead I'm thinking in a direction advocated by @waynevicknair where for the short-term we retain our existing checks for particular editors, but just clean up those checks so that we use the same approach in all cases. (For the longer-term, we should look to refactor our code (probably using object-oriented techniques) so that we can eliminate these checks.)

My current recommendation is we have our various editors attach an editor-specific property onto the editor object:
- For page editor: `editor.isPageEditor`
- For theme editor: `editor.isThemeEditor`
- For review editor: `editor.isReviewEditor`
- For CSS editor: `editor.isCSSEditor`

Then replace all occurrences of current checks for editorID or declaredClass with checks such as `if(this._editor.isPageEditor)`.

Here is my inventory of our current use of declaredClass and editorID:
- ve/States.js: subscription to "/davinci/states/state/changed" - current check is to only process if current editor is page editor. Really, this one should be fixed at the publishing level. We shouldn't use the same event for different editors, and if we had different events for different editors, we wouldn't need to check for which editor. I don't see a good way to leverage supports() in this case without making the logic unnecessarily difficult to understand. If we leave the check, better to check for declaredClass or editorID instead of using supports.
- ve/States.js" subscription to "/davinci/states/state/renamed" - we don't have rename functionality in code, so this code never gets exercised. Whatever, should have same check as previous bullet.
- ve/views/StatesView.js: subscription to "/davinci/ui/context/statesLoaded" (_statesLoaded()) - this check currently says `!themeeditor` but what it really means is `pageeditor || revieweditor`. This particular check probably could be packaged as a `supports()` call which has the meaning "this editor supports application state changes".
- ve/views/StatesView.js: subscription to "/davinci/ui/editorSelected" (_editorSelected()) - has alternate logic for themeeditor vs pageeditor.
- ve/views/StatesView.js: when user clicks on an item in Scenes/States palette, we execute different logic for theme editor vs page editor vs review editor. 
- ve/views/StatesView.js: there is a routine that hides or shows the "+" and "-" icons depending on which editor. For page editor and desktop UIs, show the icons, otherwise not.
- ve/widgets/WidgetToolBar.js: as part of editorSelected event processing, show 'id' and 'class' properties for page editor, otherwise hide those properties.
- review/editor/Context.js: subscription to "/davinci/scene/selectionChanged" - if not review editor, then return.
- review/editor/Context.js: subscription to "/davinci/states/state/changed" - if not review editor, then return.
- review/view/CommentExplorerView.js: subscription to  "/davinci/ui/editorSelected" - only process event if this is review editor.
- review/view/CommentView.js: subscription to "/davinci/states/state/changed" generated by user doc - only process event if this is review editor
- review/view/CommentView.js: subscription to "/davinci/ui/editorSelected" - different logic if new editor is review editor vs other editors
- ui/Resource.js: different logic for CSSeditor vs other editors
- ve/Context.js (2 places): A couple of logic variations for theme editor
- ve/views/SwitchingStyleView.js: Extracts the current editorID string, replaced '.' with "-", and adds a class to the top of the Properties palette. (In order to control visibility of the Common, Widget-Specific and Events panes so they don't appear for theme editor).
- ve/widgets/Cascade.js: A handful of places there is different logic for page vs theme editors.

**Comment by JonFerraiolo** [comment]
-> @JonFerraiolo M7 Prio-Med
